### ISSUE TITLE
Chapter 6 - additional peephole

### ISSUE THREAD

**Issue by dibyendumajumdar** [SeaOfNodes/Simple#40]
From Cliff (discord chat 11-Nov-23)

### Missed peephole in example 3 and 4

the test for arg is repeated twice.  Since its repeated, its the same value - true, i.e. we already tested that arg!=0 so at the second test it remains arg!=0.
The peephole is to spot this pattern, and claim the 2nd if test is always true. 
Yeah example#3 is also missing the same optimization.
This optimization does not depend on arg being known, or being a constant - which makes it much more realistic.

```
  int a = 0;
  if( arg ) {
    if( arg ) { a = 1; }
    else a = 2;// DEAD;
  }
  return a;
```

Run this code in your head.
This is a really simple optimization; don't try to out-think yourself here.
Go with your gut.
If arg is zero, a is untouched and the result is 0.
If arg is non-zero, it is so twice in a row, and only a=1 is executed.

This code is equivalent to:

```
  int a = 0;
  if( arg ) {
    a = 1;
  }
  return a;
```

Dibyendu

The change I tried was this in compute()

```java
    @Override
    public Type compute() {
        // If the If node is not reachable then neither is any following Proj
        if (ctrl()._type != Type.CONTROL) return TypeTuple.IF_NEITHER;
        // If constant is 0 then false branch is reachable
        // Else true branch is reachable
        if (pred()._type instanceof TypeInteger ti && ti.isConstant()) {
            if (ti.value() == 0)          return TypeTuple.IF_FALSE;
            else                          return TypeTuple.IF_TRUE;
        }
**        if (ctrl() instanceof ProjNode proj &&
                proj.ctrl() instanceof IfNode ifNode) {
            if (pred() == ifNode.pred()) {
                if (proj._idx == 0) return TypeTuple.IF_TRUE;
                else return TypeTuple.IF_FALSE;
            }
        }**
        return TypeTuple.IF_BOTH;
    }
```

### example 4

Right, example #4 needs control dependence.
Which we can approximate with a simple recursive algo.

Ahh, I was expecting this to be done with a idealize peephole, to collapse & remove the inner if
Your compute test works also.
Replace if (ctrl() instanceof ProjNode proj && proj.ctrl() instanceof IfNode ifNode ) with
IfNode ifNode = dominator(pred()); if( ifNode != null ) ...
And we search for the control dominator with a matching pred()

### Full Dominator Tree vs Approx approach

Cliff on discord (12-Nov-23)

My easy approx hack is maybe 10 or 20 lines.
You walk the IDOMs to the root (or cap at any arb number, C2 uses 50), looking for an If with the matching predicate.
Writing IDom tree path syntax as -*->.
If you go `{ If q -*->  T/F -> If q }`, you know the lower If is the T or F you walked through.
If you go `{ If q -*-> If q }` then you might have an unzip candidate, but these grow code, are harder to profit, and so should generally only be local, so no need for IDom.  Instead, just fail out.  Also might get `{ If q -*-> null }` or `(Start)`, also fail out.
Code is something like:

```java
 for( Node dom = idom(), prior=this; dom!=null; prior=dom, dom = dom.idom() )
    if( dom instanceof IfNode iff && iff->pred()==pred() )
      return ((ProjNode)prior)._idx==0 ? Type.CTRL : Type.XCTRL;
```

ALso, the simpler approx algo works on partial programs.  That's the point of it.
There is no way to "check that projections are on the same side", because we don't start with a projection.  We start with the lower If.  If we find a matching up If and we get to it from one side or the other - THEN we have a projection.

The incremental version looks something like this:

```java
  // Most CFG nodes:
  Node idom() { return ctrl(); }

  // Regions:
  Node RegionNode:idom() {
    if( _idom != null ) return _idom; // Return cached copy
    if( nIns()==2 ) return (_idom=in(1));
    if( nIns()> 3 ) return null; // Fails; approximate only
    Node lhs = in(1).idom();
    Node rhs = in(2).idom();
    if( lhs==null || rhs==null ) return null;
    if( lhs.idom()==rhs.idom() ) 
      return (_idom=lhs.idom());
    return null;
  }
```

Xmilia: Is it guaranteed that the _idepth of the immediate domitator is one less then the current value? I think it currently is the case but will it also hold when we add the worklist for peeps in loops? There a CFG node could be removed and will we then update the _idepth values or do we have then steps != 1?

Cliff: Yeah, without loops the invariant of +1 per level holds, because we fold extras away before computing it.
with collapsing loops we might get steps > 1
should still work as long as we have structured control flow; no n-way merges, and no irreducible loops.
n-way merges are easy to fix, but make the RegionNode somewhat more complex to look at, and aren't typically a big win here.
java gets irriduclble  loops once you have try/catch/finally in a loop that's inlined in another try/catch/finally loop - very common in large apps with retrys & time-outs around i/o and db access and the like.  So I had to deal with irreducible in a general way in C2.
Shouldn't happen here.
I know I ran this with simple algo in C2 for a long time, it might still be In There somewhere, but once I dealt with irreducible properly, the usefullness became less.

Xmilia: I saw this as a problem for the RegionNode::idom as it currently assume step size of 1 and might not find a dominator if there are other step sizes. At this point in time this is not a problem but might in the future not give the correct dominator.

Cliff: Right, this is an approx solution.  It will find a dominator not always an immediate dominator, if there's been a bunch of If's folded away after a loop head, and then the loop head folds away.  This is still correct, because the xform/peep depends on finding a dominating if, and NOT exactly the immediate dominating if. 
Humm lingo: C2 uses "xform" generically for some unspecified graph rewrite; and rarely "peephole".  Simple consistently uses "peephole", which I'm totally good with.  But sometimes I might write "xform" to mean "peephole".

Xmilia: I would prefer a loop that does allow for step > 1 such as:

```java
while( lhs!=rhs ) {
    var comp = lhs._idepth - rhs._idepth;
    if( comp >= 0 && (lhs = lhs.idom()) == null ) return null;
    if( comp <= 0 && (rhs = rhs.idom()) == null ) return null;
}
```

Cliff: 
How about:

```java
        // Walk the LHS & RHS idom trees in parallel until they match, or either fails
        Node lhs = in(1).idom();
        Node rhs = in(2).idom();
        while( lhs != rhs ) {
          if( lhs==null || rhs==null ) return null;
          var comp = lhs._idepth - rhs._idepth;
          if( comp >= 0 ) lhs = lhs.idom();
          if( comp <= 0 ) rhs = rhs.idom();
        }
        if( lhs==null ) return null;
        _idepth = lhs._idepth+1;
        return (_idom=lhs); 
```

Dibyendu: is the idepth thing a proxy for RPO?

Cliff: yes, not the RPO
Unlike the literature, this is an approx solution - it works on partial CFG's, and it works on mutating CFG's returning a conservatively correct solution.  Also, unlike the literature it can fail, and returns null instead of an idom.  So several things change:

    the idepth may differ by more than 1, but it will always be monotonic.
    the calls to prior idoms make recursively fail.

The above literature assumes that if you call doms[finger] the finger will move from less-than to equals.  Not the case for my approximation, since CFGs removed after having their idepth computed will leave number gaps.
The above literature assumes that if you call doms[finger] you'll get another finger.  Not the case again, because it can fail.

Also on the plus side, I don't need dominance frontiers, and the real code (not pretty literature psuedo-code) is pretty short. 

After some more reflection:

* the literature algo will work with idepth being monotonic, not just +1.
* my current algo will not fail for structured control flow (which we have).
* the algo WILL fail for Regions that are Loops, without special loop handling (fail is: infinite loop/crash, not a poor answer).
* But we can assume all our loops will be well structures and labeled, because we won't allow you to build irreducible loops directly from the code, and we won't do a peephole which will make an irreducible loop.

This means we can do something like the above intersect code, and skip the dominance frontier calculation. Basically, demanding a always-structured-CFG simplifies the problem and allows a simpler algorithm which gives an exact answer easier.

**Comment by dibyendumajumdar** [comment]
implemented by #42
